includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+					$instance = $jobs_post->instance;
+
+					// Populate remaining job data
+					$cron_array[ $timestamp ][ $action ][ $instance ] = array(
+						'schedule' => $jobs_post->schedule,
+						'args'     => $jobs_post->args,
+						'interval' => 0,
+					);
+
+					if ( isset( $jobs_post->interval ) ) {
+						$cron_array[ $timestamp ][ $action ][ $instance ]['interval'] = $jobs_post->interval;
+					}
+				}
+
+				// No need to keep looping if there were fewer events than we asked for
+				if ( count( $jobs_posts ) < $quantity ) {
+					break;
+				}
+			}
+		} while( true );
+
+		// Re-sort the array just as Core does when events are scheduled
+		// Ensures events are sorted chronologically
+		uksort( $cron_array, 'strnatcasecmp' );
+
+		// Cache the results, bearing in mind that they won't be used during unscheduling events
+		wp_cache_set( self::CACHE_KEY, $cron_array, null, 1 * \HOUR_IN_SECONDS );
+
+		return $cron_array;
+	}
+
+	/**
+	 * Handle requests to update the cron option
+	 *
+	 * By returning $old_value, `cron` option won't be updated
+	 */
+	public function update_option( $new_value, $old_value ) {
+		// Find changes to record
+		$new_events     = $this->find_cron_array_differences( $new_value, $old_value );
+		$deleted_events = $this->find_cron_array_differences( $old_value, $new_value );
+
+		// Add/update new events
+		foreach ( $new_events as $new_event ) {
+			$job_id = $this->job_exists( $new_event['timestamp'], $new_event['action'], $new_event['instance'], true );
+
+			if ( 0 === $job_id ) {
+				$job_id = null;
+			}
+
+			$this->create_or_update_job( $new_event['timestamp'], $new_event['action'], $new_event['args'], $job_id, false );
+		}
+
+		// Mark deleted entries for removal
+		foreach ( $deleted_events as $deleted_event ) {
+			$this->mark_job_completed( $deleted_event['timestamp'], $deleted_event['action'], $deleted_event['instance'], false );
+		}
+
+		$this->flush_internal_caches();
+
+		return $old_value;
+	}
+
+	/**
+	 * When an entry exists, don't try to create it again
+	 */
+	public function block_creation_if_job_exists( $job ) {
+		$instance = md5( maybe_serialize( $job->args ) );
+		if ( $this->job_exists( $job->timestamp, $job->hook, $instance ) ) {
+			return false;
+		}
+
+		return $job;
+	}
+
+	/**
+	 * PLUGIN UTILITY METHODS
+	 */
+
+	/**
+	 * Retrieve jobs given a set of parameters
+	 *
+	 * @param array $args
+	 * @return array|false
+	 */
+	public function get_jobs( $args ) {
+		global $wpdb;
+
+		if ( ! isset( $args['quantity'] ) || ! is_numeric( $args['quantity'] ) ) {
+			$args['quantity'] = 100;
+		}
+
+		if ( isset( $args['page'] ) ) {
+			$page  = max( 0, $args['page'] - 1 );
+			$offset = $page * $args['quantity'];
+		} else {
+			$offset = 0;
+		}
+
+		// Do not sort, otherwise index isn't used
+		$jobs = $wpdb->get_results( $wpdb->prepare( "SELECT * FROM {$this->get_table_name()} WHERE status = %s LIMIT %d,%d;", $args['status'], $offset, $args['quantity'] ), 'OBJECT' );
```

Comment:
Forgot to note that there is an index on status, care of the unique. Adding an index on status alone doesn't help, unless this query is split into one that retrieves IDs, and another that retrieves those rows using `IN`. Since it's a simple where, I'm not sure splitting is worth it.